### PR TITLE
Add tapper middleware

### PR DESCRIPTION
**Pull request by indrekj**
Enables distributed request tracing using Tapper
https://github.com/Financial-Times/tapper. The application should use
Tapper.Ctx for in process trace propagation.

screen shot 2018-08-17 at 20 59 51

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -49,10 +49,12 @@ defmodule Salemove.HttpClient.Mixfile do
       {:tesla_statsd, "~> 0.1"},
       {:confex, "~> 3.0"},
       {:poison, "~> 1.0 or ~> 2.0 or ~> 3.0"},
-      {:hackney, ">= 0.0.0", only: :dev},
+      {:hackney, ">= 0.0.0", optional: true},
```
Comment (take-five) [inline mix.exs]: Why this change?

File: mix.exs [hunk 2]
```diff
@@ -49,10 +49,12 @@ defmodule Salemove.HttpClient.Mixfile do
...
       {:ex_statsd, ">= 0.0.0", only: [:dev, :test]},
       {:ex_doc, ">= 0.0.0", only: :dev},
-      {:dialyxir, "~> 0.5", only: :dev, runtime: false}
+      {:dialyxir, "~> 0.5", only: :dev, runtime: false},
+      {:tapper, "~> 0.3", optional: true},
```
Comment (take-five) [inline mix.exs]: tapper_plug already depends on tapper

File: lib/salemove/http_client.ex [hunk 3]
```diff
@@ -138,6 +138,7 @@ defmodule Salemove.HttpClient do
     )
     |> push_middleware({Salemove.HttpClient.Middleware.Logger, options[:log]})
     |> push_middleware(Tesla.Middleware.DebugLogger, if: options[:debug])
+    |> push_middleware(Salemove.HttpClient.Middleware.Tapper, if: tapper_present?())
```
Comment (take-five) [inline lib/salemove/http_client.ex]: There should be a way to disable it. One case is when you want to send something to external system and don't want to expose internal trace IDs.

Comment (take-five) [reply lib/salemove/http_client.ex]: I mean disable it not per entire application, but per client module.

File: lib/salemove/http_client/middleware/tapper.ex [hunk 4]
```diff
@@ -0,0 +1,103 @@
+if Salemove.HttpClient.tapper_present?() do
+  defmodule Salemove.HttpClient.Middleware.Tapper do
```
Comment (take-five) [inline lib/salemove/http_client/middleware/tapper.ex]: Maybe makes sense to extract it to a separate library? It would improve ecosystem

File: lib/salemove/http_client.ex (hunk 3 above)
Comment (indrekj) [reply lib/salemove/http_client.ex]: I've seen some libraries do it in other way as well, even our rack tracer https://github.com/opentracing-contrib/ruby-rack-tracer/commit/e519eebbf6da498e896cca5fc2f466e7feee00c6#diff-3235605862566454892b34e567c5a6c6 . But I do see your point. Will add an option to disable it.

File: lib/salemove/http_client/middleware/tapper.ex (hunk 4 above)
Comment (indrekj) [reply lib/salemove/http_client/middleware/tapper.ex]: Yes makes sense. I initially started it here to get it running and integrated faster, now that I have a working solution I can extract it. Will do it soon and update this PR as well.

Comment (indrekj) [reply lib/salemove/http_client/middleware/tapper.ex]: https://github.com/salemove/elixir-tesla_middleware_tapper/pull/1 will update this pr later

File: mix.exs [hunk 5]
```diff
@@ -49,10 +49,11 @@ defmodule Salemove.HttpClient.Mixfile do
       {:tesla_statsd, "~> 0.1"},
       {:confex, "~> 3.0"},
       {:poison, "~> 1.0 or ~> 2.0 or ~> 3.0"},
-      {:hackney, ">= 0.0.0", only: :dev},
+      {:hackney, ">= 0.0.0"},
```
Comment (indrekj) [inline mix.exs]: had to change this because otherwise mix deps.get threw:
```
Dependencies have diverged:
* hackney (Hex package)
  the :only option for dependency hackney

  > In mix.exs:
    {:hackney, ">= 0.0.0", [env: :prod, repo: "hexpm", hex: "hackney", only: :dev, manager: :rebar3]}

  does not match the :only option calculated for

  > In deps/httpoison/mix.exs:
    {:hackney, "~> 1.8", [env: :prod, hex: "hackney", repo: "hexpm", optional: false]}

  Remove the :only restriction from your dep
** (Mix) Can't continue due to errors on dependencies
```

**Comment by indrekj** [comment]
@take-five updated

File: mix.exs (hunk 5 above)
Comment (take-five) [reply mix.exs]: So it became a transient dependency now? I guess we can remove it from the mix.exs completely then

File: lib/salemove/http_client.ex [hunk 6]
```diff
@@ -123,6 +123,7 @@ defmodule Salemove.HttpClient do
 
   defp build_stack(options) do
     encode_json_enabled = Keyword.get(options, :json, true)
+    enable_tapper = Keyword.get(options, :enable_tapper, true) && Code.ensure_loaded?(Tesla.Middleware.Tapper)
```
Comment (take-five) [inline lib/salemove/http_client.ex]: If it's true by default and Tapper is not included into dependencies, then we'll try to load this module on every request. I don't think that's good from performance POV and compatibility POV.
I'd rather disable it by default.

Comment (indrekj) [reply lib/salemove/http_client.ex]: hm, I thought we built the stack only once.

If this library is focused on SaleMove services (which I thought it is) then I think this should be enabled by default. I'd rather avoid adding extra steps to add tracing to our services when possible because otherwise it's very easy to forget. If the user has already installed tapper then he most likely wants to use this anyway.

Could we somehow evaluate the expression only once?

Comment (take-five) [reply lib/salemove/http_client.ex]: I guess if you make a module attribute, it'll evaluate once:
```
@tapper_enabled? Code.ensure_loaded?(Tesla.Middleware.Tapper)
```

Comment (take-five) [reply lib/salemove/http_client.ex]: In this case you can modify your check `enable_tapper = @tapper_enabled? && Keyword.get(options, :enable_tapper, true)`
The reason for this order is because `Keyword.get` has `O(N)` complexity

Comment (indrekj) [reply lib/salemove/http_client.ex]: updated
